module/mailnews/newsread.c

```diff
@@ -389,8 +412,132 @@ Bool DateFromSeconds(long seconds, char *str)
    if (t == NULL)
       return False;
 
-   sprintf(str, "%s %s %.2ld, %.4ld %.2ld:%.2ld", 
+   snprintf(str, MAXDATE, "%s %s %.2ld, %.4ld %.2ld:%.2ld", 
 	   GetString(hInst, weekdays[t->tm_wday]), GetString(hInst, months[t->tm_mon]), 
 	   t->tm_mday, t->tm_year + 1900, t->tm_hour, t->tm_min);
    return True;
 }
+
+void OnColumnClick(LPNMLISTVIEW pLVInfo)
+{
+    int nSortColumn = pLVInfo->iSubItem;
+    bool bSortAscending = TRUE;
+
+    if (nSortColumn == currentSortColumn) 
+    {
+        // Toggle sort direction if the same column is clicked again
+        currentSortAscending = !currentSortAscending;
+    } 
+    else 
+    {
+        // Set ascending order for new column
+        currentSortColumn = nSortColumn;
+        currentSortAscending = TRUE;
+    }
+
+    LPARAM lParamSort = (currentSortAscending ? 1 : -1) * (currentSortColumn + 1);
+
+    // Update header text
+    UpdateColumnHeaders(pLVInfo->hdr.hwndFrom, currentSortColumn, currentSortAscending);
+
+    // Sort list
+    ListView_SortItems(pLVInfo->hdr.hwndFrom, CompareListItems, lParamSort);
+}
+
+/****************************************************************************/
+/*
+ * CompareListItems: Comparison function for sorting items in the news dialog 
+ * list view.
+ */
+int CALLBACK CompareListItems(LPARAM lParam1, LPARAM lParam2, LPARAM lParamSort)
+{
+   bool bSortAscending = (lParamSort > 0);
+   int nColumn = abs(lParamSort) - 1;
+
+   NewsArticle *article1 = (NewsArticle *)lParam1;
+   NewsArticle *article2 = (NewsArticle *)lParam2;
+
+   // `ListView_SortItems` is unstable, so we append the article index for stability
+   char title1[MAX_SUBJECT + 10];
+   snprintf(title1, sizeof(title1), "%s %d", article1->title, article1->num);
+   char title2[MAX_SUBJECT + 10];
+   snprintf(title2, sizeof(title2), "%s %d", article2->title, article2->num);
+
+   char poster1[MAXUSERNAME + 10];
+   snprintf(poster1, sizeof(poster1), "%s %d", article1->poster, article1->num);
+   char poster2[MAXUSERNAME + 10];
+   snprintf(poster2, sizeof(poster2), "%s %d", article2->poster, article2->num);
+
+   switch (nColumn)
+   {
+   case COL_TITLE:
+      return bSortAscending ? stricmp(title1, title2) : stricmp(title2, title1);
+   case COL_POSTER:
+      return bSortAscending ? stricmp(poster1, poster2) : stricmp(poster2, poster1);
+   case COL_TIME:
+      return bSortAscending ? article1->time - article2->time : article2->time - article1->time;
+   }
+
+   return 0;
```

Comment:
This would be a good place for an assertion or error print, as the switch should always handle all cases, so this code is unreachable.

Reply:
If it is reached, and it's an assertion or error print, the callback continues repeatedly down the list and produces either many assertion dialogs or redundant error messages for each comparison that takes place. From what I understand, we can't break from the comparison prematurely, so maybe this is what it is? This is what I have currently:

```
switch (nColumn)
{
   case COL_TITLE:
      return bSortAscending ? stricmp(title1, title2) : stricmp(title2, title1);
   case COL_POSTER:
      return bSortAscending ? stricmp(poster1, poster2) : stricmp(poster2, poster1);
   case COL_TIME:
      return bSortAscending ? article1->time - article2->time : article2->time - article1->time;
   default:
      debug(("Unhandled column (column #%d) in CompareListItems\n", nColumn));
      return 0;
}
```

Reply:
That's fine.  Hitting that print should really "never happen", as it would mean a coding error, not a runtime error.